feat!: Customizable Slashing and Jailing

stana-miric opened the pull request:
Closes https://github.com/cosmos/interchain-security/issues/2383

This PR introduces the ability for each consumer chain to set custom infraction parameters for slashing and jailing (double signing and downtime). The MsgCreateConsumer and MsgUpdateConsumer messages have been extended to support this functionality.

MsgCreateConsumer: 
If infraction parameters are not provided in the message, the values will default to those from the provider and will be stored per consumer chain id.

MsgUpdateConsumer:
        Pre-launch phase: Parameters are updated immediately.
        Post-launch phase: Parameters are added to a queue, and the update will take effect after the unbonding period expires.

Queue Logic:
    If parameters for the consumer already exist in the queue:
       If the new parameters match the current infraction param for that consumer, the existing queued entry is removed to cancel the update. If the new parameters differ, the existing queued entry is replaced with the new parameters.
    If parameters do not exist in the queue and the new parameters differ from the current ones, a new entry is added to schedule the update.

The unbonding period expiration is checked in the BeginBlocker, leveraging the existing time queue infrastructure.
When a consumer chain is removed, related queue records will also be deleted, but the infraction parameters will remain( following the same logic applied for power parameters). Additionally, a migration script has been implemented to set default infraction parameters for all currently active consumer chains.

mpoke commented on .changelog/unreleased/features/provider/2403-customizable-slashing-jailing.md:
This is empty. Also, move outside `provider` folder. Just add `[x/provider]` at the beginning of the entry.

stana-miric replied:
fixed [69ec878](https://github.com/cosmos/interchain-security/pull/2403/commits/69ec878bb060642cd82ed11f3a9d7c8a04d2ebf3)

mpoke commented on .changelog/unreleased/state-breaking/2403-customizable-slashing-jailing.md:
```diff
@@ -0,0 +1,2 @@
+- Allow consumer chains to customize the slashing and jailing conditions. Every consumer chain can decide the punishment for every type of infraction.
```
```suggestion
- `[x/provider]` Enable the customization of the slashing and jailing conditions for infractions committed by validators on consumer chains (as per [ADR 020](https://cosmos.github.io/interchain-security/adrs/adr-020-cutomizable_slashing_and_jailing)). Every consumer chain can decide the punishment for every type of infraction.
```

stana-miric replied:
fixed [69ec878](https://github.com/cosmos/interchain-security/pull/2403/commits/69ec878bb060642cd82ed11f3a9d7c8a04d2ebf3)

mpoke commented on docs/docs/features/slashing.md:
```diff
@@ -14,8 +14,8 @@ The ICS protocol differentiates between downtime and equivocation infractions.
 Downtime infractions are reported by consumer chains and are acted upon on the provider as soon as they are received. 
-Instead of slashing, the provider will **_only jail_** offending validator for the duration of time established by the provider chain parameters.
-Note that validators are only jailed for downtime on consumer chains that they opted in to validate on,
+The provider will jail and slash the offending validator. The jailing duration and slashing fraction are determined by the consumer's downtime infraction parameters on the provider chain.
+Note that validators are only slashed and jailed for downtime on consumer chains that they opted in to validate on,
 or in the case of Top N chains, where they are automatically opted in by being in the Top N% of the validator set on the provider.
```
```suggestion
The provider will jail and slash the offending validator. The jailing duration and slashing fraction are determined by the consumer's downtime infraction parameters on the provider chain.
By default, validators are **_only jailed_** for downtime on consumer chains that they opted in to validate on,
or in the case of Top N chains, where they are automatically opted in by being in the Top N% of the validator set on the provider.
```

stana-miric replied:
fixed [69ec878](https://github.com/cosmos/interchain-security/pull/2403/commits/69ec878bb060642cd82ed11f3a9d7c8a04d2ebf3)

mpoke commented on docs/docs/features/slashing.md:
Please also update https://cosmos.github.io/interchain-security/build/modules/provider, i.e., state, messages, client

stana-miric replied:
fixed [69ec878](https://github.com/cosmos/interchain-security/pull/2403/commits/69ec878bb060642cd82ed11f3a9d7c8a04d2ebf3)

mpoke commented on proto/interchain_security/ccv/provider/v1/provider.proto:
```diff
@@ -556,3 +556,22 @@ enum ConsumerPhase {
+    (amino.dont_omitempty) = true
+  ];
+  // for permanent jailing use 9223372036854775807 which is the largest value a time.Duration can hold (approximately 292 years)
+  google.protobuf.Duration jail_duration = 8
```
```suggestion
  google.protobuf.Duration jail_duration = 2
```

stana-miric replied:
fixed [69ec878](https://github.com/cosmos/interchain-security/pull/2403/commits/69ec878bb060642cd82ed11f3a9d7c8a04d2ebf3)

mpoke commented on x/ccv/provider/migrations/v9/migrations.go:
This should be added in the upgrade handler of the provider chain (i.e., Gaia). This is because the migration can be done using the provider keeper. So add this logic in UPGRADING.md. Also, remove the consensus version bump from module.go.

mpoke replied:
You could also open a PR on Gaia that bumps ICS to this version and adds the migration to the upgrade handler.

stana-miric replied:
not sure under which section in UPGRADING.md to add the code for upgrade handles. I guess under "Unreleased" section but is seems like some sections under Unreleased are not up to date ?

I will open the PR, but I guess I should to that later? "bumps ICS to this version " which exact ics version to use?

mpoke replied:
Update first the info in UPGRADING.md using https://github.com/cosmos/interchain-security/blob/v6.3.0/UPGRADING.md. Then add the migration under Unreleased. 

> I will open the PR, but I guess I should to that later? "bumps ICS to this version " which exact ics version to use?

Sure. Or just use a commit. The important thing is to not forget about the upgrade handler in Gaia.

stana-miric replied:
Fixed [2b0eb15](https://github.com/cosmos/interchain-security/pull/2403/commits/2b0eb1576c510161da3564374344c56e6af9d467)

https://github.com/cosmos/gaia/pull/3430

mpoke reviewed (commented):
Partial review. See comments below.

mpoke reviewed (commented):
Approve the general logic. Great work @stana-miric. See my comments below. 

Please add also interchain tests to ensure that updating the infraction params works.

stana-miric commented on x/ccv/provider/keeper/relay.go:
```diff
@@ -478,6 +478,20 @@ func (k Keeper) HandleSlashPacket(ctx sdk.Context, consumerId string, data ccv.S
+	if err != nil {
+		k.Logger(ctx).Error("failed to slash vaidator", providerConsAddr.ToSdkConsAddr().String(), "err", err.Error())
+		return
+	}
```
fixed [69ec878](https://github.com/cosmos/interchain-security/pull/2403/commits/69ec878bb060642cd82ed11f3a9d7c8a04d2ebf3)

stana-miric commented:
> Approve the general logic. Great work @stana-miric. See my comments below.
> 
> Please add also interchain tests to ensure that updating the infraction params works.

Cannot add interchain tests at this point cuz its another module and I can import the right version of ics in ics interchain test only after this pr is merged. I've added test that tests: adding infraction parameters through MsgCreateConsumer and MsgUpdateConsumer, also added tests that test queuing and overriding existing queued item as well as cancelling update, and tests that confirms that when jailing and lashing is called for misbehavior on consumer chain that parameters for that specific consumer is used.
I was thisnking to test this with interchain with the consumer suite that we're currently working on and will use those custon infraction parameters for consume chains.

mpoke commented:
> > Approve the general logic. Great work @stana-miric. See my comments below.
> > Please add also interchain tests to ensure that updating the infraction params works.
> 
> Cannot add interchain tests at this point cuz its another module and I can import the right version of ics in ics interchain test only after this pr is merged. 

Sounds good. 

> I've added test that tests: adding infraction parameters through MsgCreateConsumer and MsgUpdateConsumer, also added tests that test queuing and overriding existing queued item as well as cancelling update, and tests that confirms that when jailing and lashing is called for misbehavior on consumer chain that parameters for that specific consumer is used. 

Great 

> I was thisnking to test this with interchain with the consumer suite that we're currently working on and will use those custon infraction parameters for consume chains.

Not sure we need to consumer side. That needed in general to test infractions (downtime and double signing). This feature just enables consumer chain owners to change the params. So the e2e test for it should just involve the provider API -- create consumer and update consumer. For example, can we use interchain tests to verify that an update will take the unbonding period? without waiting for 3 weeks :)

mpoke reviewed (approved):
Great work @stana-miric

stana-miric commented:
> > > Approve the general logic. Great work @stana-miric. See my comments below.
> > > Please add also interchain tests to ensure that updating the infraction params works.
> > 
> > 
> > Cannot add interchain tests at this point cuz its another module and I can import the right version of ics in ics interchain test only after this pr is merged.
> 
> Sounds good.
> 
> > I've added test that tests: adding infraction parameters through MsgCreateConsumer and MsgUpdateConsumer, also added tests that test queuing and overriding existing queued item as well as cancelling update, and tests that confirms that when jailing and lashing is called for misbehavior on consumer chain that parameters for that specific consumer is used.
> 
> Great
> 
> > I was thisnking to test this with interchain with the consumer suite that we're currently working on and will use those custon infraction parameters for consume chains.
> 
> Not sure we need to consumer side. That needed in general to test infractions (downtime and double signing). This feature just enables consumer chain owners to change the params. So the e2e test for it should just involve the provider API -- create consumer and update consumer. For example, can we use interchain tests to verify that an update will take the unbonding period? without waiting for 3 weeks :)

Interchain tests are now implemented [here](https://github.com/cosmos/interchain-security/pull/2416). Once this pr is merged, pr with tests can be updated(ics version in go.mod, and image version) and reviewed

MSalopek commented on proto/interchain_security/ccv/provider/v1/tx.proto:
```diff
@@ -363,6 +363,9 @@ message MsgCreateConsumer {
   AllowlistedRewardDenoms allowlisted_reward_denoms = 6;
+
+  // infraction parameters for slashing and jailing
+  InfractionParameters infraction_parameters = 7;
```
Is this optional when the Tx is submitted via CLI/RPC?

It seems very annoying to think about such details during MsgCreateConsumer.

stana-miric replied:
yes, its optional. if its not set, default values(same as the provider values) will be set.

MSalopek reviewed (approved):
LGTM.

I still need to do an deep dive into the tests. The logic actually seems straightforward - any time a slash/jail operation needs to take place we first consult the consumer parameters and apply them instead of using the sdk defaults.

stana-miric commented on UPGRADING.md:
```diff
@@ -1,7 +1,43 @@
 
 ## Unreleased
 
+### Provider
```
fixed [a068fcb](https://github.com/cosmos/interchain-security/pull/2403/commits/a068fcb0a434234de16cdd8eb69849a14690290a)
